usaspending_api/download/filestreaming/csv_generation.py

```diff
@@ -0,0 +1,284 @@
+
+
+class CsvSource:
+    def __init__(self, model_type, file_type, source_type):
+        self.model_type = model_type
+        self.file_type = file_type
+        self.source_type = source_type
+        self.human_names = download_column_historical_lookups.human_names[model_type][file_type]
+        self.query_paths = download_column_historical_lookups.query_paths[model_type][file_type]
+        self.queryset = None
+
+    def values(self, header):
+        query_paths = [self.query_paths[hn] for hn in header]
+        return self.queryset.values_list(query_paths).iterator()
+
+    def columns(self, requested):
+        """Given a list of column names requested, returns the ones available in the source"""
+        result = self.human_names
+        if requested:
+            result = [header for header in requested if header in self.human_names]
+
+        # remove headers that we don't have a query path for
+        result = [header for header in result if header in self.query_paths]
+
+        return result
+
+    def row_emitter(self, headers_requested):
+        headers = self.columns(headers_requested)
+        # Not yielding headers as the files can be split
+        # yield headers
+        query_paths = [self.query_paths[hn] for hn in headers]
+        return self.queryset.values(*query_paths)
+
+
+def generate_csvs(download_job, sqs_message=None):
+    """Derive the relevant file location and write CSVs to it"""
+    start_time = time.time()
+
+    # Parse data from download_job
+    json_request = json.loads(download_job.json_request)
+    columns = json_request.get('columns', None)
+    limit = json_request.get('limit', None)
+
+    file_name = start_download(download_job)
+    try:
+        # Create temporary files and working directory
+        file_path = settings.BULK_DOWNLOAD_LOCAL_PATH + file_name
+        working_dir = os.path.splitext(file_path)[0]
+        if not os.path.exists(working_dir):
+            os.mkdir(working_dir)
+        zipped_csvs = zipfile.ZipFile(file_path, 'w', allowZip64=True)
+
+        write_to_log(message='Generating {}'.format(file_name), download_job=download_job)
+
+        # Generate sources from the JSON request object
+        sources = get_csv_sources(json_request)
+        for source in sources:
+            # Parse and write data to the file
+            download_job.number_of_columns = max(download_job.number_of_columns, len(source.columns(columns)))
+            parse_source(source, columns, download_job, working_dir, start_time, sqs_message, zipped_csvs, limit)
+
+        # Remove temporary files and working directory
+        shutil.rmtree(working_dir)
+        zipped_csvs.close()
+        download_job.file_size = os.stat(file_path).st_size
+    except Exception as e:
+        logger.error(e)
+        handle_file_generation_exception(file_path, download_job, 'write', str(e))
+
+    try:
+        # push file to S3 bucket, if not local
+        if not settings.IS_LOCAL:
```

Comment:
If we're not doing a local download do we want to delete the local file or is this done somewhere that I can't find?

Reply:
Added deletion